# Notebook to pack Zhang68k data from x10genomics into an AnnData (h5ad) file.  

This example follows [Zheng]() for identification of white blood cell types from single cell RNA expression data.


## Outline of process
The finetune process requrires the input data to be in AnnDatam (h5ad) format with the cell types
as labels. If the data is not packed as AnnData, as is the case when downloading from the
10xgenomics site as explained below, it need to first be packed into one and saved to the disk.

Once the data is in anndata format, it needs to be canonified via data and variable name transformations.
After the tranformations the filtered and cannonified data may be saved again in anndata format or used directly.  This data can be used by the training process.

In the training process, each sample is transforemed into a GeneFormer like ordered list of genes, which is done by sorting all the genes by binned expression level and within each bin alphabetically by the gene name (as it is used in the toenizer).
This is done inside the task.process_model_output method, together with input preperations and tokenizations.






## preprocessing data transformations

There is a commented example of this process at the end of this notebook.  

* Filter out cell with less then 200 active genes

        scanpy.pp.filter_cells(anndata_object,min_genes=200)



* Normelize the sum of counts for each cell to a constant (1000)

        scanpy.pp.normalize_total(anndata_object,1000.)


* Move to log space (note,the data prior to this step will be in the range 0-10 so scanpy may issue a warning that the data has allready beed log-scaled)

        scanpy.pp.log1p(anndata_object,base=2)


* split the full range of values into bins and digitize the values

        bins=np.linspace(anndata_object.X.data.min(), anndata_object.X.max(),num=10)
        anndata_object.X.data=np.digitize(anndata_object.X.data, bins)

    Note that this was done over all the data, but that is not likely to cause any bleeding from the test sets



This notebook assumes that it is ran from the `bmfm-mammal-release/mammal/examples/scrna_cell_type` directory.  

In [ ]:
# check the current directory. 
# Notice that the `biomed-multi-alignment` will probably be placed in a different location on your system.

# this is just a fancy pwd, replace with !pwd if you need
!print -D $PWD

~/git/biomed-multi-alignment/mammal/examples/scrna_cell_type/data



## Obtaining the source data:
The main data is availble online, for example in the [10xgenomics](https://www.10xgenomics.com/) cite.  The lables are based on the data in [LINK](https://www.10xgenomics.com/datasets/fresh-68-k-pbm-cs-donor-a-1-standard-1-1-0)

From this download the file `fresh_68k_pbmc_donor_a_filtered_gene_bc_matrices.tar.gz` and place it in this directory. 

In [ ]:
!ls -sh

total 2052152
     8 README.md
917616 Zhang_68k.h5ad
886152 Zhang_68k_processed.h5ad
     0 filtered_matrices_mex
243056 fresh_68k_pbmc_donor_a_filtered_gene_bc_matrices.tar.gz
     8 process_h5ad_data.py
    32 zhang_data_prep.ipynb
  2640 zheng17_bulk_lables.txt
  2640 zheng17_bulk_lables.txt.1


#### Unzip the file.

In [ ]:
!tar -xzvf fresh_68k_pbmc_donor_a_filtered_gene_bc_matrices.tar.gz 

x filtered_matrices_mex/
x filtered_matrices_mex/hg19/
x filtered_matrices_mex/hg19/barcodes.tsv
x filtered_matrices_mex/hg19/genes.tsv
x filtered_matrices_mex/hg19/matrix.mtx


#### download the labels file from a git repository in https://github.com/scverse/scanpy_usage

In [ ]:
!wget https://raw.githubusercontent.com/scverse/scanpy_usage/refs/heads/master/170503_zheng17/data/zheng17_bulk_lables.txt

--2025-02-19 12:29:45--  https://raw.githubusercontent.com/scverse/scanpy_usage/refs/heads/master/170503_zheng17/data/zheng17_bulk_lables.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8001::154, 2606:50c0:8002::154, 2606:50c0:8003::154, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8001::154|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1348927 (1.3M) [text/plain]
Saving to: ‘zheng17_bulk_lables.txt.2’

zheng17_bulk_lables 100%[===================>]   1.29M   940KB/s    in 1.4s    

2025-02-19 12:29:47 (940 KB/s) - ‘zheng17_bulk_lables.txt.2’ saved [1348927/1348927]



  You should now have a directoy called `filtered_matrices_mex/hg19` 

In [ ]:
!ls -shR --color=never


total 2056288
     8 README.md
917616 Zhang_68k.h5ad
886152 Zhang_68k_processed.h5ad
     0 filtered_matrices_mex
243056 fresh_68k_pbmc_donor_a_filtered_gene_bc_matrices.tar.gz
     8 process_h5ad_data.py
    32 zhang_data_prep.ipynb
  2640 zheng17_bulk_lables.txt
  2640 zheng17_bulk_lables.txt.1
  4136 zheng17_bulk_lables.txt.2

./filtered_matrices_mex:
total 0
0 hg19

./filtered_matrices_mex/hg19:
total 993144
  2280 barcodes.tsv	  1600 genes.tsv	989264 matrix.mtx


The output should be something like

```
total 245728
     8 README.md
     0 filtered_matrices_mex
243056 fresh_68k_pbmc_donor_a_filtered_gene_bc_matrices.tar.gz
    24 zhang_data_prep.ipynb
  2640 zheng17_bulk_lables.txt

./filtered_matrices_mex:
total 0
0 hg19

./filtered_matrices_mex/hg19:
total 993144
  2280 barcodes.tsv	  1600 genes.tsv	989264 matrix.mtx

## Pack the data into an AnnData file 

In [ ]:
import anndata 
# from collections import Counter
import pandas as pd
# import matplotlib.pyplot as plt
# import math
import numpy as np
# from scipy.sparse import csr_matrix
from scipy.io import mmread
import scanpy as sc

#### Read the scRNA matrix from a file


In [ ]:
mmx = mmread("filtered_matrices_mex/hg19/matrix.mtx")


#### Create an AnnData object wrapping the read data

Notice that this code transposes the data to the correct direction

In [ ]:
 
anndata_object = anndata.AnnData(X=mmx.transpose().tocsr())
print(anndata_object.X.shape)

(68579, 32738)


In [ ]:
# Cell identifiers
barcodes = pd.read_csv("filtered_matrices_mex/hg19/barcodes.tsv",header=None,sep="\t")
# names of genes
genes = pd.read_csv("filtered_matrices_mex/hg19/genes.tsv",header=None,sep="\t")


In [ ]:
# cell types (this is actualy just one column)
cell_type_lables = pd.read_csv("zheng17_bulk_lables.txt",header=None,sep="\t")

In [ ]:

# use the gene names as variable names in the AnnData object
anndata_object.var_names=genes[1]

# use the cell barcodes as names for the samples
anndata_object.obs_names=barcodes[0]



In [ ]:

# use cell types as labels for the samples
anndata_object.obs['celltype']=cell_type_lables.squeeze().to_numpy()

In [ ]:
# Save result anndata object to disk
anndata_object.write_h5ad("Zhang_68k.h5ad")

### And the annData file is ready in the data directory.

---------


In [ ]:
# # process the data - filter out cells with shallow reads, normelize depth and change to log scale of about 0-10 (log_2(1001)~=10)

# sc.pp.filter_cells(anndata_object,min_genes=200)
# sc.pp.normalize_total(anndata_object,1000.)
# sc.pp.log1p(anndata_object,base=2)


# anndata_object.X.data.min(), anndata_object.X.max()
# # split range to bins - more or less 0-1,1-2,2-3,..10
# bins=np.linspace(anndata_object.X.data.min(), anndata_object.X.max(),num=11)
# print(bins)


# # convert the counts to bins
# anndata_object.X.data=np.digitize(anndata_object.X.data, bins)

# # Save result anndata object to disk
# anndata_object.write_h5ad("Zhang68k_filtered.h5ad")



# inside the dataset pipeline:
# def convert_to_double_sorted_geneformer_sequance(anndata_object):

#     # the genes are sorted by expression bin (decending) and within the bin by the gene names.    
#     return [a[1] for a in sorted(zip(-anndata_object.X.data,anndata_object.var_names.to_numpy()[anndata_object.X.indices]))]
